### ISSUE TITLE
Package request: Mermaid and mmdc

### ISSUE THREAD

**Issue by matthew-piziak** [NixOS/nixpkgs#60057]
[Mermaid](https://mermaidjs.github.io/) generates diagrams and flowcharts from text in a similar manner as markdown. Typically the required component is `mmdc`, which is the binary that generates images from the markup representation.

GitLab [natively generates](https://docs.gitlab.com/ee/user/markdown.html#mermaid) Mermaid diagrams. Emacs has [a mode](https://github.com/abrochard/mermaid-mode/tree/06beffbbadc4ed25d22f82d377e92f47d79588fd) for it that supports local rendering. Support for GitHub does not exist yet but is [highly demanded](https://github.com/github/markup/issues/533) and is seeing some upstream support. Being able to install it on NixOS would be useful.

**Comment by matthew-piziak** [comment]
I'd be tempted to add it to `nodePackages` (see https://www.npmjs.com/package/mermaid), but my [last attempted contribution](https://github.com/NixOS/nixpkgs/pull/54443) has been stuck in purgatory for _four months_, so I'm not exactly eager to try it until that one is resolved.

**Comment by danbst** [comment]
@matthew-piziak I became interested in having mermaid as well

**Comment by danbst** [comment]
@matthew-piziak I was able to build mermaid.cli off-tree.

Create a `node-packages.json` (do that preferably in separate directory called `mermaid`):
```
[
    "mermaid",
    "mermaid.cli"
]
```
Run `nix-shell -p nodePackages.node2nix --nodejs10 -i node-packages.json` to generate `default.nix`, `node-packages.nix` and `node-env.nix`.

Then create a wrapper package:
```
let  = import ./mermaid/default.nix { };
in mermaidCli.overrideAttrs (sup: {
            nativeBuildInputs = sup.nativeBuildInputs or [] ++ [
                nixpkgs.makeWrapper nixpkgs.which
            ];

            PUPPETEER_SKIP_CHROMIUM_DOWNLOAD = 1;

            nixpkgsChromePuppeteerConfig = nixpkgs.writeText "puppeteerConfig.json" ''
                { "executablePath": "${chromium}/bin/chromium" }
            '';

            postInstall = sup.postInstall or "" + ''
                mv $out/bin/mmdc $out/bin/mmdc-script
                NODE=$(readlink -f $(which node))
                cat > $out/bin/mmdc <<EOF
                #!/usr/bin/env bash
                exec $NODE $out/bin/mmdc-script -p $nixpkgsChromePuppeteerConfig "\$@"
                EOF
                chmod +x $out/bin/mmdc
            '';
});
```
This was enough to work with `asciidoctor-diagrams`.

**Comment by matthew-piziak** [comment]
Hm. I get `error: attribute 'nodejs-8_x' missing` in `mermaid/default.nix` when I try this.

**Comment by matthew-piziak** [comment]
This started failing for me with `request to http://www.example.com/chalk failed, reason: getaddrinfo ENOTFOUND www.example.com www.example.com:80` so I removed it from my config.

**Comment by ysndr** [comment]
Took another shot to include it in nixpkgs here: #87766. It works well on MacOSwith the test file I used.

**Comment by ryneeverett** [comment]
Resolved via #87766.
